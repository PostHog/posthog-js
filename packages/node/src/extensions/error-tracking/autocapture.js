// Portions of this file are derived from getsentry/sentry-javascript by Software, Inc. dba Sentry
// Licensed under the MIT License
function makeUncaughtExceptionHandler(captureFn, onFatalFn) {
    let calledFatalError = false;
    return Object.assign((error) => {
        // Attaching a listener to `uncaughtException` will prevent the node process from exiting. We generally do not
        // want to alter this behaviour so we check for other listeners that users may have attached themselves and adjust
        // exit behaviour of the SDK accordingly:
        // - If other listeners are attached, do not exit.
        // - If the only listener attached is ours, exit.
        const userProvidedListenersCount = global.process.listeners('uncaughtException').filter((listener) => {
            // There are 2 listeners we ignore:
            return (
            // as soon as we're using domains this listener is attached by node itself
            listener.name !== 'domainUncaughtExceptionClear' &&
                // the handler we register in this integration
                listener._posthogErrorHandler !== true);
        }).length;
        const processWouldExit = userProvidedListenersCount === 0;
        captureFn(error, {
            mechanism: {
                type: 'onuncaughtexception',
                handled: false,
            },
        });
        if (!calledFatalError && processWouldExit) {
            calledFatalError = true;
            onFatalFn();
        }
    }, { _posthogErrorHandler: true });
}
export function addUncaughtExceptionListener(captureFn, onFatalFn) {
    global.process.on('uncaughtException', makeUncaughtExceptionHandler(captureFn, onFatalFn));
}
export function addUnhandledRejectionListener(captureFn) {
    global.process.on('unhandledRejection', (reason) => {
        captureFn(reason, {
            mechanism: {
                type: 'onunhandledrejection',
                handled: false,
            },
        });
    });
}
//# sourceMappingURL=autocapture.js.map