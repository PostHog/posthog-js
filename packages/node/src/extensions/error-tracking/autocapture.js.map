{"version":3,"file":"autocapture.js","sourceRoot":"","sources":["autocapture.ts"],"names":[],"mappings":"AAAA,kGAAkG;AAClG,iCAAiC;AAMjC,SAAS,4BAA4B,CACnC,SAAsD,EACtD,SAAqB;IAErB,IAAI,gBAAgB,GAAY,KAAK,CAAA;IAErC,OAAO,MAAM,CAAC,MAAM,CAClB,CAAC,KAAY,EAAQ,EAAE;QACrB,8GAA8G;QAC9G,kHAAkH;QAClH,yCAAyC;QACzC,kDAAkD;QAClD,iDAAiD;QACjD,MAAM,0BAA0B,GAAG,MAAM,CAAC,OAAO,CAAC,SAAS,CAAC,mBAAmB,CAAC,CAAC,MAAM,CAAC,CAAC,QAAQ,EAAE,EAAE;YACnG,mCAAmC;YACnC,OAAO;YACL,0EAA0E;YAC1E,QAAQ,CAAC,IAAI,KAAK,8BAA8B;gBAChD,8CAA8C;gBAC7C,QAAyB,CAAC,oBAAoB,KAAK,IAAI,CACzD,CAAA;QACH,CAAC,CAAC,CAAC,MAAM,CAAA;QAET,MAAM,gBAAgB,GAAG,0BAA0B,KAAK,CAAC,CAAA;QAEzD,SAAS,CAAC,KAAK,EAAE;YACf,SAAS,EAAE;gBACT,IAAI,EAAE,qBAAqB;gBAC3B,OAAO,EAAE,KAAK;aACf;SACF,CAAC,CAAA;QAEF,IAAI,CAAC,gBAAgB,IAAI,gBAAgB,EAAE,CAAC;YAC1C,gBAAgB,GAAG,IAAI,CAAA;YACvB,SAAS,EAAE,CAAA;QACb,CAAC;IACH,CAAC,EACD,EAAE,oBAAoB,EAAE,IAAI,EAAE,CAC/B,CAAA;AACH,CAAC;AAED,MAAM,UAAU,4BAA4B,CAC1C,SAAsD,EACtD,SAAqB;IAErB,MAAM,CAAC,OAAO,CAAC,EAAE,CAAC,mBAAmB,EAAE,4BAA4B,CAAC,SAAS,EAAE,SAAS,CAAC,CAAC,CAAA;AAC5F,CAAC;AAED,MAAM,UAAU,6BAA6B,CAAC,SAAwD;IACpG,MAAM,CAAC,OAAO,CAAC,EAAE,CAAC,oBAAoB,EAAE,CAAC,MAAe,EAAE,EAAE;QAC1D,SAAS,CAAC,MAAM,EAAE;YAChB,SAAS,EAAE;gBACT,IAAI,EAAE,sBAAsB;gBAC5B,OAAO,EAAE,KAAK;aACf;SACF,CAAC,CAAA;IACJ,CAAC,CAAC,CAAA;AACJ,CAAC"}