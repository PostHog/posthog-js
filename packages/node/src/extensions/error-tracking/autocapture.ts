// Portions of this file are derived from getsentry/sentry-javascript by Software, Inc. dba Sentry
// Licensed under the MIT License

import { EventHint } from './types'

type ErrorHandler = { _posthogErrorHandler: boolean } & ((error: Error) => void)

function makeUncaughtExceptionHandler(
  captureFn: (exception: Error, hint: EventHint) => void,
  onFatalFn: () => void
): ErrorHandler {
  let calledFatalError: boolean = false

  return Object.assign(
    (error: Error): void => {
      // Attaching a listener to `uncaughtException` will prevent the node process from exiting. We generally do not
      // want to alter this behaviour so we check for other listeners that users may have attached themselves and adjust
      // exit behaviour of the SDK accordingly:
      // - If other listeners are attached, do not exit.
      // - If the only listener attached is ours, exit.
      const userProvidedListenersCount = global.process.listeners('uncaughtException').filter((listener) => {
        // There are 2 listeners we ignore:
        return (
          // as soon as we're using domains this listener is attached by node itself
          listener.name !== 'domainUncaughtExceptionClear' &&
          // the handler we register in this integration
          (listener as ErrorHandler)._posthogErrorHandler !== true
        )
      }).length

      const processWouldExit = userProvidedListenersCount === 0

      captureFn(error, {
        mechanism: {
          type: 'onuncaughtexception',
          handled: false,
        },
      })

      if (!calledFatalError && processWouldExit) {
        calledFatalError = true
        onFatalFn()
      }
    },
    { _posthogErrorHandler: true }
  )
}

export function addUncaughtExceptionListener(
  captureFn: (exception: Error, hint: EventHint) => void,
  onFatalFn: () => void
): void {
  global.process.on('uncaughtException', makeUncaughtExceptionHandler(captureFn, onFatalFn))
}

export function addUnhandledRejectionListener(captureFn: (exception: unknown, hint: EventHint) => void): void {
  global.process.on('unhandledRejection', (reason: unknown) => {
    captureFn(reason, {
      mechanism: {
        type: 'onunhandledrejection',
        handled: false,
      },
    })
  })
}
