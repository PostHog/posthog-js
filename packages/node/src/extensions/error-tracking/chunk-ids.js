// Portions of this file are derived from getsentry/sentry-javascript by Software, Inc. dba Sentry
// Licensed under the MIT License
let parsedStackResults;
let lastKeysCount;
let cachedFilenameChunkIds;
export function getFilenameToChunkIdMap(stackParser) {
    const chunkIdMap = globalThis._posthogChunkIds;
    if (!chunkIdMap) {
        console.error('No chunk id map found');
        return {};
    }
    const chunkIdKeys = Object.keys(chunkIdMap);
    if (cachedFilenameChunkIds && chunkIdKeys.length === lastKeysCount) {
        return cachedFilenameChunkIds;
    }
    lastKeysCount = chunkIdKeys.length;
    cachedFilenameChunkIds = chunkIdKeys.reduce((acc, stackKey) => {
        if (!parsedStackResults) {
            parsedStackResults = {};
        }
        const result = parsedStackResults[stackKey];
        if (result) {
            acc[result[0]] = result[1];
        }
        else {
            const parsedStack = stackParser(stackKey);
            for (let i = parsedStack.length - 1; i >= 0; i--) {
                const stackFrame = parsedStack[i];
                const filename = stackFrame?.filename;
                const chunkId = chunkIdMap[stackKey];
                if (filename && chunkId) {
                    acc[filename] = chunkId;
                    parsedStackResults[stackKey] = [filename, chunkId];
                    break;
                }
            }
        }
        return acc;
    }, {});
    return cachedFilenameChunkIds;
}
//# sourceMappingURL=chunk-ids.js.map