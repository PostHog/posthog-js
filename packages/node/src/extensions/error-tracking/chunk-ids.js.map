{"version":3,"file":"chunk-ids.js","sourceRoot":"","sources":["chunk-ids.ts"],"names":[],"mappings":"AAAA,kGAAkG;AAClG,iCAAiC;AASjC,IAAI,kBAAiE,CAAA;AACrE,IAAI,aAAiC,CAAA;AACrC,IAAI,sBAAkD,CAAA;AAEtD,MAAM,UAAU,uBAAuB,CAAC,WAAwB;IAC9D,MAAM,UAAU,GAAI,UAAkB,CAAC,gBAA8C,CAAA;IACrF,IAAI,CAAC,UAAU,EAAE,CAAC;QAChB,OAAO,CAAC,KAAK,CAAC,uBAAuB,CAAC,CAAA;QACtC,OAAO,EAAE,CAAA;IACX,CAAC;IAED,MAAM,WAAW,GAAG,MAAM,CAAC,IAAI,CAAC,UAAU,CAAC,CAAA;IAE3C,IAAI,sBAAsB,IAAI,WAAW,CAAC,MAAM,KAAK,aAAa,EAAE,CAAC;QACnE,OAAO,sBAAsB,CAAA;IAC/B,CAAC;IAED,aAAa,GAAG,WAAW,CAAC,MAAM,CAAA;IAElC,sBAAsB,GAAG,WAAW,CAAC,MAAM,CAAyB,CAAC,GAAG,EAAE,QAAQ,EAAE,EAAE;QACpF,IAAI,CAAC,kBAAkB,EAAE,CAAC;YACxB,kBAAkB,GAAG,EAAE,CAAA;QACzB,CAAC;QAED,MAAM,MAAM,GAAG,kBAAkB,CAAC,QAAQ,CAAC,CAAA;QAE3C,IAAI,MAAM,EAAE,CAAC;YACX,GAAG,CAAC,MAAM,CAAC,CAAC,CAAC,CAAC,GAAG,MAAM,CAAC,CAAC,CAAC,CAAA;QAC5B,CAAC;aAAM,CAAC;YACN,MAAM,WAAW,GAAG,WAAW,CAAC,QAAQ,CAAC,CAAA;YAEzC,KAAK,IAAI,CAAC,GAAG,WAAW,CAAC,MAAM,GAAG,CAAC,EAAE,CAAC,IAAI,CAAC,EAAE,CAAC,EAAE,EAAE,CAAC;gBACjD,MAAM,UAAU,GAAG,WAAW,CAAC,CAAC,CAAC,CAAA;gBACjC,MAAM,QAAQ,GAAG,UAAU,EAAE,QAAQ,CAAA;gBACrC,MAAM,OAAO,GAAG,UAAU,CAAC,QAAQ,CAAC,CAAA;gBAEpC,IAAI,QAAQ,IAAI,OAAO,EAAE,CAAC;oBACxB,GAAG,CAAC,QAAQ,CAAC,GAAG,OAAO,CAAA;oBACvB,kBAAkB,CAAC,QAAQ,CAAC,GAAG,CAAC,QAAQ,EAAE,OAAO,CAAC,CAAA;oBAClD,MAAK;gBACP,CAAC;YACH,CAAC;QACH,CAAC;QAED,OAAO,GAAG,CAAA;IACZ,CAAC,EAAE,EAAE,CAAC,CAAA;IAEN,OAAO,sBAAsB,CAAA;AAC/B,CAAC"}