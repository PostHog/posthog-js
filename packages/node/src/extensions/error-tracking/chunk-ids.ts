// Portions of this file are derived from getsentry/sentry-javascript by Software, Inc. dba Sentry
// Licensed under the MIT License

import type { StackParser } from './types'

type StackString = string
type CachedResult = [string, string]

type ChunkIdMapType = Record<string, string>

let parsedStackResults: Record<StackString, CachedResult> | undefined
let lastKeysCount: number | undefined
let cachedFilenameChunkIds: ChunkIdMapType | undefined

export function getFilenameToChunkIdMap(stackParser: StackParser): ChunkIdMapType {
  const chunkIdMap = (globalThis as any)._posthogChunkIds as ChunkIdMapType | undefined
  if (!chunkIdMap) {
    console.error('No chunk id map found')
    return {}
  }

  const chunkIdKeys = Object.keys(chunkIdMap)

  if (cachedFilenameChunkIds && chunkIdKeys.length === lastKeysCount) {
    return cachedFilenameChunkIds
  }

  lastKeysCount = chunkIdKeys.length

  cachedFilenameChunkIds = chunkIdKeys.reduce<Record<string, string>>((acc, stackKey) => {
    if (!parsedStackResults) {
      parsedStackResults = {}
    }

    const result = parsedStackResults[stackKey]

    if (result) {
      acc[result[0]] = result[1]
    } else {
      const parsedStack = stackParser(stackKey)

      for (let i = parsedStack.length - 1; i >= 0; i--) {
        const stackFrame = parsedStack[i]
        const filename = stackFrame?.filename
        const chunkId = chunkIdMap[stackKey]

        if (filename && chunkId) {
          acc[filename] = chunkId
          parsedStackResults[stackKey] = [filename, chunkId]
          break
        }
      }
    }

    return acc
  }, {})

  return cachedFilenameChunkIds
}
