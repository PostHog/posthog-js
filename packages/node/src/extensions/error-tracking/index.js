import { addUncaughtExceptionListener, addUnhandledRejectionListener } from './autocapture';
import { uuidv7 } from '@posthog/core/vendor/uuidv7';
import { propertiesFromUnknownInput } from './error-conversion';
const SHUTDOWN_TIMEOUT = 2000;
export default class ErrorTracking {
    static async buildEventMessage(error, hint, distinctId, additionalProperties) {
        const properties = { ...additionalProperties };
        // Given stateless nature of Node SDK we capture exceptions using personless processing when no
        // user can be determined because a distinct_id is not provided e.g. exception autocapture
        if (!distinctId) {
            properties.$process_person_profile = false;
        }
        const exceptionProperties = await propertiesFromUnknownInput(this.stackParser, this.frameModifiers, error, hint);
        return {
            event: '$exception',
            distinctId: distinctId || uuidv7(),
            properties: {
                ...exceptionProperties,
                ...properties,
            },
        };
    }
    constructor(client, options) {
        this.client = client;
        this._exceptionAutocaptureEnabled = options.enableExceptionAutocapture || false;
        this.startAutocaptureIfEnabled();
    }
    startAutocaptureIfEnabled() {
        if (this.isEnabled()) {
            addUncaughtExceptionListener(this.onException.bind(this), this.onFatalError.bind(this));
            addUnhandledRejectionListener(this.onException.bind(this));
        }
    }
    onException(exception, hint) {
        void ErrorTracking.buildEventMessage(exception, hint).then((msg) => {
            this.client.capture(msg);
        });
    }
    async onFatalError() {
        await this.client.shutdown(SHUTDOWN_TIMEOUT);
    }
    isEnabled() {
        return !this.client.isDisabled && this._exceptionAutocaptureEnabled;
    }
}
//# sourceMappingURL=index.js.map