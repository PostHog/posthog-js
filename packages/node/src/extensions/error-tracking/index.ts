import { EventHint, StackFrameModifierFn, StackParser } from './types'
import { addUncaughtExceptionListener, addUnhandledRejectionListener } from './autocapture'
import { PostHogBackendClient } from '../../client'
import { uuidv7 } from 'posthog-core/src/vendor/uuidv7'
import { propertiesFromUnknownInput } from './error-conversion'
import { EventMessage, PostHogOptions } from '../../types'

const SHUTDOWN_TIMEOUT = 2000

export default class ErrorTracking {
  private client: PostHogBackendClient
  private _exceptionAutocaptureEnabled: boolean

  static stackParser: StackParser
  static frameModifiers: StackFrameModifierFn[]

  static async captureException(
    client: PostHogBackendClient,
    error: unknown,
    hint: EventHint,
    distinctId?: string,
    additionalProperties?: Record<string | number, any>
  ): Promise<void> {
    const properties: EventMessage['properties'] = { ...additionalProperties }

    // Given stateless nature of Node SDK we capture exceptions using personless processing when no
    // user can be determined because a distinct_id is not provided e.g. exception autocapture
    if (!distinctId) {
      properties.$process_person_profile = false
    }

    const exceptionProperties = await propertiesFromUnknownInput(this.stackParser, this.frameModifiers, error, hint)

    client.capture({
      event: '$exception',
      distinctId: distinctId || uuidv7(),
      properties: {
        ...exceptionProperties,
        ...properties,
      },
    })
  }

  constructor(client: PostHogBackendClient, options: PostHogOptions) {
    this.client = client
    this._exceptionAutocaptureEnabled = options.enableExceptionAutocapture || false

    this.startAutocaptureIfEnabled()
  }

  private startAutocaptureIfEnabled(): void {
    if (this.isEnabled()) {
      addUncaughtExceptionListener(this.onException.bind(this), this.onFatalError.bind(this))
      addUnhandledRejectionListener(this.onException.bind(this))
    }
  }

  private onException(exception: unknown, hint: EventHint): void {
    ErrorTracking.captureException(this.client, exception, hint)
  }

  private async onFatalError(): Promise<void> {
    await this.client.shutdown(SHUTDOWN_TIMEOUT)
  }

  isEnabled(): boolean {
    return !this.client.isDisabled && this._exceptionAutocaptureEnabled
  }
}
