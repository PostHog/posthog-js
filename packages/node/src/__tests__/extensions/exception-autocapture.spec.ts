import ErrorTracking from '@/extensions/error-tracking'
import { PostHog } from '@/entrypoints/index.node'
import { addUncaughtExceptionListener, addUnhandledRejectionListener } from '@/extensions/error-tracking/autocapture'
import { Worker } from 'worker_threads'

describe('exception autocapture', () => {
  it('should capture uncaught exception', () => {
    global.process.on = jest.fn()
    addUncaughtExceptionListener(
      () => {},
      () => {}
    )
    expect(global.process.on).toHaveBeenCalledWith('uncaughtException', expect.any(Function))
  })

  it('should capture unhandled rejection', () => {
    global.process.on = jest.fn()
    addUnhandledRejectionListener(() => {})
    expect(global.process.on).toHaveBeenCalledWith('unhandledRejection', expect.any(Function))
  })

  it('should listen to uncaught errors', async () => {
    const worker = new Worker(__dirname + '/exception-autocapture.worker.mjs')
    const capturePromise = new Promise<void>((res, rej) => {
      worker.on('message', (message) => {
        expect(message.method).toBe('capture')
        expect(message.event).toMatchSnapshot()
        res()
      })
    })
    worker.postMessage({ action: 'throw_error', data: 'Uncaught Error' })
    await capturePromise
  })

  it('should listen to unhandled rejections', async () => {
    const worker = new Worker(__dirname + '/exception-autocapture.worker.mjs')
    const capturePromise = new Promise<void>((res, rej) => {
      worker.on('message', (message) => {
        expect(message.method).toBe('capture')
        expect(message.event).toMatchSnapshot()
        res()
      })
    })
    worker.postMessage({ action: 'reject_promise', data: 'Unhandled Promise' })
    await capturePromise
  })

  it('should rate limit when more than 10 of the same exception are caught', async () => {
    jest.spyOn(ErrorTracking, 'buildEventMessage').mockResolvedValue({
      event: '$exception',
      distinctId: 'distinct-id',
      properties: { $exception_list: [{ type: 'Error' }] },
    })

    const ph = new PostHog('TEST_API_KEY', {
      host: 'http://example.com',
      fetchRetryCount: 0,
      disableCompression: true,
    })

    const mockedCapture = jest.spyOn(ph, 'capture').mockImplementation()

    const captureExceptions = Array.from({ length: 20 }).map(() => ph['errorTracking']['onException']({}, {}))
    await Promise.all(captureExceptions)

    // captures until rate limited
    expect(mockedCapture).toHaveBeenCalledTimes(9)
  })
})
