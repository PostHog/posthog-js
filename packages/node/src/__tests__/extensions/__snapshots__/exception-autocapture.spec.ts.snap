// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exception autocapture should listen to uncaught errors 1`] = `
{
  "distinctId": "stable_id",
  "event": "$exception",
  "properties": {
    "$exception_level": "error",
    "$exception_list": [
      {
        "mechanism": {
          "handled": false,
          "synthetic": false,
          "type": "onuncaughtexception",
        },
        "stacktrace": {
          "frames": [
            {
              "colno": 28,
              "filename": "node:internal/per_context/messageport",
              "function": "MessagePort.?",
              "in_app": false,
              "lineno": 23,
              "module": "messageport",
              "platform": "node:javascript",
            },
            {
              "colno": 20,
              "filename": "node:internal/event_target",
              "function": "[nodejs.internal.kHybridDispatch]",
              "in_app": false,
              "lineno": 831,
              "module": "event_target",
              "platform": "node:javascript",
            },
            {
              "colno": 11,
              "context_line": "    throw new Error(msg.data)",
              "filename": "/Users/hpouillot/Code/posthog-js/packages/node/src/__tests__/extensions/exception-autocapture.worker.mjs",
              "function": "MessagePort.?",
              "in_app": true,
              "lineno": 10,
              "module": "exception-autocapture.worker",
              "platform": "node:javascript",
              "post_context": [
                "  } else if (msg.action == 'reject_promise') {",
                "    return Promise.reject(new Error(msg.data))",
                "  } else {",
                "    console.error('Unrecognized message from main thread:', msg)",
                "  }",
                "})",
                "",
              ],
              "pre_context": [
                "",
                "const posthog = new PostHog('api_key', {",
                "  enableExceptionAutocapture: true,",
                "})",
                "",
                "parentPort.on('message', (msg) => {",
                "  if (msg.action == 'throw_error') {",
              ],
            },
          ],
          "type": "raw",
        },
        "type": "Error",
        "value": "Uncaught Error",
      },
    ],
    "$process_person_profile": false,
  },
}
`;

exports[`exception autocapture should listen to unhandled rejections 1`] = `
{
  "distinctId": "stable_id",
  "event": "$exception",
  "properties": {
    "$exception_level": "error",
    "$exception_list": [
      {
        "mechanism": {
          "handled": false,
          "synthetic": false,
          "type": "onuncaughtexception",
        },
        "stacktrace": {
          "frames": [
            {
              "colno": 28,
              "filename": "node:internal/per_context/messageport",
              "function": "MessagePort.?",
              "in_app": false,
              "lineno": 23,
              "module": "messageport",
              "platform": "node:javascript",
            },
            {
              "colno": 20,
              "filename": "node:internal/event_target",
              "function": "[nodejs.internal.kHybridDispatch]",
              "in_app": false,
              "lineno": 831,
              "module": "event_target",
              "platform": "node:javascript",
            },
            {
              "colno": 27,
              "context_line": "    return Promise.reject(new Error(msg.data))",
              "filename": "/Users/hpouillot/Code/posthog-js/packages/node/src/__tests__/extensions/exception-autocapture.worker.mjs",
              "function": "MessagePort.?",
              "in_app": true,
              "lineno": 12,
              "module": "exception-autocapture.worker",
              "platform": "node:javascript",
              "post_context": [
                "  } else {",
                "    console.error('Unrecognized message from main thread:', msg)",
                "  }",
                "})",
                "",
                "await new Promise((res) => {",
                "  posthog.capture = (event) => {",
              ],
              "pre_context": [
                "  enableExceptionAutocapture: true,",
                "})",
                "",
                "parentPort.on('message', (msg) => {",
                "  if (msg.action == 'throw_error') {",
                "    throw new Error(msg.data)",
                "  } else if (msg.action == 'reject_promise') {",
              ],
            },
          ],
          "type": "raw",
        },
        "type": "Error",
        "value": "Unhandled Promise",
      },
    ],
    "$process_person_profile": false,
  },
}
`;
